### PR TITLE
Extract deferred jobs to a separate process

### PR DESCRIPTION
**Pull request by jafrog**
The main loop that handles subscription to the queue and publishing messages to a normal queue once they-re ready is left intact.

What changed is that instead of a thread this loop runs in a process. It runs infinitely, but we catch interrupts and ensure that channel is closed.

On the publishing side `QueueManager` now implements it's own `publish_deferred` method so it doesn't depend on `DeferredWorkManager`.

### CODE REVIEW

File: bin/deferred.rb [hunk 1]
```diff
@@ -0,0 +1,10 @@
+require_relative "../lib/tom_queue.rb"
```
Comment (jafrog) [inline bin/deferred.rb]: This file is just for showcasing. We should delete it and replace with a real script to launch the process

File: lib/tom_queue/deferred_work_manager.rb [hunk 2]
```diff
@@ -20,130 +24,36 @@ class DeferredWorkManager
...
-    # Returns nil
-    def self.reset!
-      @@singletons_mutex.synchronize do
-        @@singletons.each_pair do |k,v|
-          v.ensure_stopped
-        end
-        @@singletons = {}
-      end
-      nil
-    end
```
Comment (jafrog) [inline lib/tom_queue/deferred_work_manager.rb]: I removed everything related to singletons. The idea is that it's just one `DeferredWorkManager` instance per process

File: lib/tom_queue/deferred_work_set.rb [hunk 3]
```diff
@@ -87,7 +89,9 @@ def pop(timeout)
 
           begin
             end_time = [next_run_at, timeout_end].compact.min
+            debug "End Time: #{end_time}"
```
Comment (jafrog) [inline lib/tom_queue/deferred_work_set.rb]: TODO: remove

File: lib/tom_queue/deferred_work_set.rb [hunk 4]
```diff
@@ -87,7 +89,9 @@ def pop(timeout)
...
             delay = end_time - Time.now
+            debug "Delay: #{delay}"
```
Comment (jafrog) [inline lib/tom_queue/deferred_work_set.rb]: TODO: remove

File: lib/tom_queue/deferred_work_set.rb [hunk 5]
```diff
@@ -8,6 +8,8 @@ module TomQueue
   #
   class DeferredWorkSet
 
+    include TomQueue::LoggingHelper
```
Comment (jafrog) [inline lib/tom_queue/deferred_work_set.rb]: TODO: remove

File: lib/tom_queue/deferred_work_set.rb [hunk 6]
```diff
@@ -8,6 +8,8 @@ module TomQueue
   #
   class DeferredWorkSet
```
Comment (jafrog) [inline lib/tom_queue/deferred_work_set.rb]: I didn't change any mutex and condvar related logic in this class. Can it be removed?

File: lib/tom_queue/deferred_work_manager.rb (hunk 2 above)
Comment (joshado) [inline lib/tom_queue/deferred_work_manager.rb]: Yep that should work.

File: lib/tom_queue/deferred_work_set.rb (hunk 6 above)
Comment (joshado) [inline lib/tom_queue/deferred_work_set.rb]: I wouldn't want to - lets just make the smallest change to break out the deferred handler.

File: lib/tom_queue/deferred_work_manager.rb [hunk 7]
```diff
@@ -163,69 +63,50 @@ def setup_amqp(channel)
...
+      error e
       reporter = TomQueue.exception_reporter
       reporter && reporter.notify($!)
+    end
 
-    ensure
-      @channel && @channel.close
-      @deferred_set = nil
-      @thread = nil      
+    def stop
```
Comment (joshado) [inline lib/tom_queue/deferred_work_manager.rb]: I'm not sure of the ordering of this, vs. the main loop, given you're in an exception handler. I tend to do the minimum possible work in such handlers.

I'd suggest setting a `@shutdown` flag to true, then calling `deferred_set && deferred_set.interrupt` to make the main-thread loop spin once and hit the `until @shutdown` to break out. 

Then the two `consumer.cancel` and `channel.close` can live after the loop itself has exited.

**Pull request merged**